### ISSUE TITLE
Meeting outcomes 10/09/03

### ISSUE THREAD

**Issue by amartinhuertas** [GiteonCaulfied/COMP4560_stokes_ml_project#3]
We were discussing about which problem to solve, and the details underlying the generation of a training dataset. Discussions are still in a preliminary stage, so do not take the following as set in stone. Comments and edits of the below are welcome ...

@sghelichkhani ... please review and add as per required.

We have come up with an step-0  problem, with the following features:

* Transient linear mantle convection on a square (lets keep aside non-linearities temporarily)
* Fixed mesh, of, e.g., 150x150 quadrilaterals.
* Inf-Sup stable Finite Elements for the spatial discretization.
* Crank-Nicolson (Implicit) time discretization. Time step size? Fixed or adaptive?
* Constant parameter values, i.e., no spatially-varying viscosity, etc. In a first stage, we can generate a dataset with these parameter values fixed. In a more advanced stage, with applicability in mind, they will have to become input parameters to the ML system as well.
* Initial condition (temperature) is not constant, but a spatially-varying field. It becomes the main input parameter to the ML system, and the variable to sample in order to generate an initial dataset. 
* Fixed time domain $[0,T]$. Value of $T$?

The first ML system should ideally be able to predict the temperature field at time $t_{n+1}$ given the temperature field at time $t_{n}$, with $n=1,\ldots,N$,  and $N$ the number of parts in which we discretize $[0,T]$. This would be like a sort of ML-based cheap Forward Stokes solver. 

As we do not actually know if this can be even be achieved (perhaps too much ambitious), we said that we can define a set of predefined points in time (say, after, 10%, 25%, 50%, 75% of time elapsed) in which we are going to observe the mapping that transforms the solution in the next time step from the previous one, and then train different neural networks for each of these moments in time. Then, we can evaluate their accuracy in predicting unseen simulations at these moments in time. 

In my view, CNNs (convolutional neural networks) would be the method of choice for such task. The issue that we have now is how should we feed the CNNs, either with the raw field values resulting from the numerical solver (read from HDF5 files), or a rasterization into 2D RGB images otherwise. The conclusion is that we should do some extra research to determine this, although I am relatively certain that usually CNNs work with images. See, e.g, the following paper https://www.sciencedirect.com/science/article/pii/S0009250921004516. They say:

> The input images are ten sequential frames with a size of 64 x 64 x 3, which are concatenated into one tensor of size 64 x 64 x 30, and the output is one frame with the size of 64 x 64 x 3.

BTW, in this paper, they are packing the first 10x time frames of the simulation into a single tensor, and then predict the output at the next time step. 

Some additional concerns:
* Should we train the ML system with temperature only data, or may it make sense to also feed it with velocities and pressures?
* Is the numerical solver robust regardless of the physical parameter values? (this is important when we start sampling these). To reduce noice, we could restrict ourselves to sparse direct solvers at each time step.
* What else?
